### ISSUE TITLE
Declared type preference can suppress invalid specialization diagnostics

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#4301]
For example, the first call correctly produces two diagnostics due to the incompatible upper bounds inferred from the callable types, while the second call suppresses the diagnostics, as [the declared type preference](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/src/types/call/bind.rs#L6024) is not validated against the inferred upper bound.

```py
from typing import Any, Callable

def f[T](
    value: T,
    first: Callable[[T], None],
    second: Callable[[T], None],
) -> list[T]:
    return [value]

def _(value: Any, ints: Callable[[int], None], strings: Callable[[str], None]):
    # error: [invalid-argument-type]
    # error: [invalid-argument-type]
    f(value, ints, strings)

    # ok
    x: list[Any] = f(value, ints, strings)
```
